Improve `--check-cfg` implementation

Urgau opened the pull request:
This pull-request is a mix of improvements regarding the `--check-cfg` implementation:

- Simpler internal representation (usage of `Option` instead of separate bool)
- Add --check-cfg to the unstable book (based on the RFC)
- Improved diagnostics:
    * List possible values when the value is unexpected
    * Suggest if possible a name or value that is similar
- Add more tests (well known names, mix of combinations, ...)

r? @petrochenkov

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5275889190?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/aa754d7902fdbb9c91f8595aaca81cab33457478/checks/5275889190/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......................................................i.ii......................................... 12600/12660
............................................................
failures:

---- [ui] ui/check-cfg/mix.rs stdout ----


1 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:10:7
+   --> $DIR/mix.rs:11:7
3    |
4 LL | #[cfg(widnows)]
5    |       ^^^^^^^ help: did you mean: `windows`
7    = note: `#[warn(unexpected_cfgs)]` on by default
8 
9 warning: unexpected `cfg` condition value
-   --> $DIR/mix.rs:20:7
-   --> $DIR/mix.rs:20:7
+   --> $DIR/mix.rs:21:7
11    |
12 LL | #[cfg(feature = "zebra")]


15    = note: possible values for `feature` are: bar, foo
16 
17 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:24:12
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+   --> $DIR/mix.rs:25:12
19    |
20 LL | #[cfg_attr(uu, test)]

22 
22 
23 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:33:10
+   --> $DIR/mix.rs:34:10
25    |
26 LL |     cfg!(widnows);
27    |          ^^^^^^^ help: did you mean: `windows`
28 
29 warning: unexpected `cfg` condition value
-   --> $DIR/mix.rs:37:10
+   --> $DIR/mix.rs:38:10
+   --> $DIR/mix.rs:38:10
31    |
32 LL |     cfg!(feature = "zebra");


35    = note: possible values for `feature` are: bar, foo
36 
37 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:39:10
+   --> $DIR/mix.rs:40:10
39    |
40 LL |     cfg!(xxx = "foo");

42 
42 
43 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:41:10
+   --> $DIR/mix.rs:42:10
45    |
46 LL |     cfg!(xxx);

48 
48 
49 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:43:23
+   --> $DIR/mix.rs:44:23
51    |
52 LL |     cfg!(any(windows, xxx));

54 
54 
55 warning: unexpected `cfg` condition name
-   --> $DIR/mix.rs:45:14
+   --> $DIR/mix.rs:46:14
57    |
58 LL |     cfg!(any(xxx, windows));

60 
61 warning: unexpected `cfg` condition value
-   --> $DIR/mix.rs:47:14
-   --> $DIR/mix.rs:47:14
+   --> $DIR/mix.rs:48:14
63    |
64 LL |     cfg!(any(feature = "bad", windows));


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/check-cfg/mix/mix.stderr
To only update this specific test, also pass `--test-args check-cfg/mix.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/check-cfg/mix.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/check-cfg/mix" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--check-cfg=names()" "--check-cfg=values(feature,\"foo\")" "--cfg" "feature=\"bar\"" "-Z" "unstable-options" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/check-cfg/mix/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:11:7
   |
LL | #[cfg(widnows)]
   |       ^^^^^^^ help: did you mean: `windows`
   = note: `#[warn(unexpected_cfgs)]` on by default

warning: unexpected `cfg` condition value
  --> /checkout/src/test/ui/check-cfg/mix.rs:21:7
  --> /checkout/src/test/ui/check-cfg/mix.rs:21:7
   |
LL | #[cfg(feature = "zebra")]
   |
   |
   = note: possible values for `feature` are: bar, foo

warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:25:12
   |
LL | #[cfg_attr(uu, test)]


warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:34:10
   |
LL |     cfg!(widnows);
   |          ^^^^^^^ help: did you mean: `windows`
warning: unexpected `cfg` condition value
  --> /checkout/src/test/ui/check-cfg/mix.rs:38:10
   |
   |
LL |     cfg!(feature = "zebra");
   |
   |
   = note: possible values for `feature` are: bar, foo

warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:40:10
   |
LL |     cfg!(xxx = "foo");


warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:42:10
   |
LL |     cfg!(xxx);


warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:44:23
   |
LL |     cfg!(any(windows, xxx));


warning: unexpected `cfg` condition name
  --> /checkout/src/test/ui/check-cfg/mix.rs:46:14
   |
LL |     cfg!(any(xxx, windows));

warning: unexpected `cfg` condition value
  --> /checkout/src/test/ui/check-cfg/mix.rs:48:14
   |
   |
LL |     cfg!(any(feature = "bad", windows));
   |                        |
   |                        |
   |                        help: did you mean: `"bar"`
   |
   = note: possible values for `feature` are: bar, foo
warning: 10 warnings emitted


------------------------------------------
```

</details>

petrochenkov commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -463,27 +464,40 @@ pub fn cfg_matches(cfg: &ast::MetaItem, sess: &ParseSess, features: Option<&Feat
             MetaItemKind::NameValue(..) | MetaItemKind::Word => {
                 let name = cfg.ident().expect("multi-segment cfg predicate").name;
```
```suggestion
                let ident = cfg.ident().expect("multi-segment cfg predicate");
```
Well, the span is used now.

petrochenkov commented on compiler/rustc_interface/src/interface.rs:
```diff
@@ -219,7 +225,11 @@ pub fn parse_check_cfg(specs: Vec<String>) -> CheckCfg {
 
-        cfg.names_valid.extend(cfg.values_checked.iter().cloned());
+        if let Some(values_valid) = &cfg.values_valid {
+            if let Some(names_valid) = &mut cfg.names_valid {
```
Can be merged into a single `if let`.

Urgau replied:
I tried that at first, but it's now yet complete in beta.
```rust
error[E0658]: `let` expressions in this position are experimental
   --> compiler/rustc_interface/src/interface.rs:228:12
    |
228 |         if let Some(values_valid) = &cfg.values_valid && let Some(names_valid) = &mut cfg.names_valid {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information
    = help: add `#![feature(let_chains)]` to the crate attributes to enable
    = help: you can write `matches!(<expr>, <pattern>)` instead of `let <pattern> = <expr>`
```
Then:
```rust
error: the feature `let_chains` is incomplete and may not be safe to use and/or cause compiler crashes
 --> compiler/rustc_interface/src/lib.rs:4:12
  |
4 | #![feature(let_chains)]
  |            ^^^^^^^^^^
  |
  = note: `-D incomplete-features` implied by `-D warnings`
  = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

```

Urgau replied:
Oh! Addressed in https://github.com/rust-lang/rust/pull/94175/commits/7de5f3b815f0146d2c964e625d2a99d89361aeb5.

petrochenkov commented on compiler/rustc_lint/src/context.rs:
```diff
@@ -766,7 +766,39 @@ pub trait LintContext: Sized {
+                BuiltinLintDiagnostics::UnexpectedCfg(span, name, value) => {
+                    let mut possibilities: Vec<Symbol> = if value.is_some() {
+                        let Some(values_valid) = &sess.parse_sess.check_config.values_valid else {
+                            bug!("it shouldn't be possible to have a diagnostic on a value if values checking is not enable");
```
```suggestion
                            bug!("it shouldn't be possible to have a diagnostic on a value if value checking is not enabled");
```

petrochenkov commented on compiler/rustc_lint/src/context.rs:
```diff
@@ -766,7 +766,39 @@ pub trait LintContext: Sized {
+                        values.iter().map(|&s| s).collect()
+                    } else {
+                        let Some(names_valid) = &sess.parse_sess.check_config.names_valid else {
+                            bug!("it shouldn't be possible to have a diagnostic on a value if values checking is not enable");
```
```suggestion
                            bug!("it shouldn't be possible to have a diagnostic on a value if value checking is not enabled");
```

petrochenkov commented on compiler/rustc_lint/src/context.rs:
```diff
@@ -766,7 +766,39 @@ pub trait LintContext: Sized {
+                    // for names as the possibilities could be very long
+                    if value.is_some() && !possibilities.is_empty() {
+                        // Sorting can take some time, so we only do it if required
+                        possibilities.sort();
```
Sorting `Symbol`s is not very reproducible, it's better to sort their string representations.

Urgau replied:
I wanted to avoid creating another `Vec` but you're right it's better and safer.

petrochenkov commented on compiler/rustc_lint/src/context.rs:
```diff
@@ -766,7 +766,39 @@ pub trait LintContext: Sized {
+                        possibilities.sort();
+
+                        let possibilities = possibilities.iter().map(Symbol::as_str).intersperse(", ").collect::<String>();
+                        db.note(&format!("possible values for `{name}` are: {possibilities}"));
```
```suggestion
                        db.note(&format!("expected values for `{name}` are: {possibilities}"));
```

petrochenkov commented on compiler/rustc_lint/src/context.rs:
```diff
@@ -766,7 +766,39 @@ pub trait LintContext: Sized {
+
+                    // Suggest the most probable if we found one
+                    if let Some(best_match) = find_best_match_for_name(&possibilities, value.unwrap_or(name), None) {
+                        let ponctuation = if value.is_some() { "\"" } else { "" };
```
```suggestion
                        let punctuation = if value.is_some() { "\"" } else { "" };
```

petrochenkov commented on compiler/rustc_session/src/config.rs:
```diff
@@ -1023,34 +1023,28 @@ pub fn to_crate_config(cfg: FxHashSet<(String, Option<String>)>) -> CrateConfig
+    /// The set of all `names()`, if none no names checking is performed
+    pub names_valid: Option<FxHashSet<T>>,
+    /// The set of all `values()`, if none no values chcking is performed
+    pub values_valid: Option<FxHashMap<T, FxHashSet<T>>>,
```
I don't think this is correct, for values there's no single `Option`.
`--check-cfg values(foo, ...)` enables value checking only for `foo` but not for other names.

petrochenkov replied:
Ok, I see, if values are not checked for the name `bar`, then `bar` is simply not in the map, so it doesn't affect correctness.

But I still think the `Option` is redundant there, because there's no observable difference between `None` and `Some(<empty map>)`.

>It's also probably faster to check `None` than trying to get a element from an empty `FxHashMap`

https://github.com/rust-lang/hashbrown/pull/305

Urgau replied:
> Ok, I see, if values are not checked for the name `bar`, then `bar` is simply not in the map, so it doesn't affect correctness.

Yes exactly, but being in it without wanting to be in would be a problem because not having values is a perfectly valid state, ex: `unix`, `windows`, ... this would affect correctness.

> But I still think the `Option` is redundant there, because there's no observable difference between `None` and `Some(<empty map>)`.

Fair enough, done in https://github.com/rust-lang/rust/pull/94175/commits/106bac126e126959311b9504ea08b80dc39c39eb.

Urgau commented on src/doc/unstable-book/src/compiler-flags/check-cfg.md:
```diff
@@ -0,0 +1,221 @@
+
+## The `names(...)` form
+
+This form uses a named metadata list:
```
I don't know, I took this directly from the RFC. I've rephrase this.

Urgau commented on src/doc/unstable-book/src/compiler-flags/check-cfg.md:
```diff
@@ -0,0 +1,221 @@
+
+If `--check-cfg names(...)` is specified at least once, then `rustc` will check all references to
+condition names. `rustc` will check every `#[cfg]` attribute, `#[cfg_attr]` attribute, and
+`cfg!(...)` call against the provided list of expected condition names. If a name is not present in
```
I didn't knew, thks. Added.

Urgau commented:
I've address all review comments, except for https://github.com/rust-lang/rust/pull/94175#discussion_r812064579 and https://github.com/rust-lang/rust/pull/94175#discussion_r812044898 where I left a comment.
This is ready for a another review.

I also would like to apologies for my English mistakes, it's not my native language. I will try my best to make sure these mistakes doesn't happen again.

@rustbot ready

Urgau commented:
Fixed https://github.com/rust-lang/rust/pull/94175#discussion_r811934197 and added a note when no value is expected https://github.com/rust-lang/rust/pull/94175/commits/ecd3a659113aff97d924517593eeed560664cf66.

@rustbot ready

petrochenkov commented on compiler/rustc_session/src/config.rs:
```diff
@@ -1023,34 +1023,30 @@ pub fn to_crate_config(cfg: FxHashSet<(String, Option<String>)>) -> CrateConfig
-    pub values_valid: FxHashSet<(T, T)>,
+    /// The set of all `names()`, if none no names checking is performed
+    pub names_valid: Option<FxHashSet<T>>,
+    /// The set of all `values()`, if none no values chcking is performed
```
Outdated comment about none.

petrochenkov commented on compiler/rustc_interface/src/interface.rs:
```diff
@@ -183,13 +184,16 @@ pub fn parse_check_cfg(specs: Vec<String>) -> CheckCfg {
-                                        cfg.values_checked.insert(ident.to_string());
+                                        let ident_values = cfg
+                                            .values_valid
+                                            .entry(ident.to_string())
```
`ident.to_string()` is too smart and is supposed to only be used in diagnostics (see how `impl fmt::Display for Ident` is implemented).
`ident.name`/`ident.name.to_string()` needs to be used here instead.

petrochenkov commented:
r=me with the remaining comments addressed and commits squashed appropriately.

Urgau commented:
Fixed last comments and squashed to 5 commits. This is ready for r=you.

@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 6e3f642eaa7778306aafd08c4f6f7c24bdb176b5 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 6e3f642eaa7778306aafd08c4f6f7c24bdb176b5 -->
<!-- homu: {"type":"Approved","sha":"6e3f642eaa7778306aafd08c4f6f7c24bdb176b5","approver":"petrochenkov"} -->

matthiaskrgr commented:
@bors r- failed in a rollup
https://github.com/rust-lang/rust/pull/94281#issuecomment-1048682357

Urgau commented:
The failure is due to the eager evaluation of cfg condition done with `any()`. I have for now remove the problematic test case.

Note: That I started modifying the code to always evaluate every cfg condition in `any()` but ran into a breaking change where `aa:bb` in `#[cfg(any(unix, aa::bb))]` would only produce a error if `unix` evaluate to false. [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5ff97acda4bb6a4a563e80083410686c). I don't think that a problem given the unusual condition required for triggering the bug but that a problem for another PR.

@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit a556a2a8e60501203f310407b27cf739618c0000 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov a556a2a8e60501203f310407b27cf739618c0000 -->
<!-- homu: {"type":"Approved","sha":"a556a2a8e60501203f310407b27cf739618c0000","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit a556a2a8e60501203f310407b27cf739618c0000 with merge d0a34dc40eba51c4cd4173f8fb56b4c562209c90...
<!-- homu: {"type":"BuildStarted","head_sha":"a556a2a8e60501203f310407b27cf739618c0000","merge_sha":"d0a34dc40eba51c4cd4173f8fb56b4c562209c90"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5322224750?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5322224750?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5322185028?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/d0a34dc40eba51c4cd4173f8fb56b4c562209c90/checks/5322185028/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>
